### PR TITLE
BAAS-24348: estimate mem usage for gomap and goslice when over thresholds

### PR DESCRIPTION
**Pull request by Gabri3l**
This is something we already do for arrays and objects so I ported over the logic. I did some renaming just to be consistent with other mem usage functions (I don't feel strongly about it though).

### CODE REVIEW

File: object_goslice.go [hunk 1]
```diff
@@ -350,24 +350,69 @@ func (o *objectGoSlice) swap(i int, j int) {
...
+	var samplesVisited, memUsage, newMemUsage uint64
+	counter := 0
+	sliceLen := len(*o.data)
+	sampleSize := sliceLen / 10
+
+	// grabbing one sample every "sampleSize" to provide consistent
+	// memory usage across function executions
+	for _, datum := range *o.data {
+		counter++
+		if counter%sampleSize == 0 {
```
Comment (nathantfrank) [inline object_goslice.go:366]: [minor] similar to other comment.

File: object_gomap.go [hunk 2]
```diff
@@ -157,23 +157,68 @@ func (o *objectGoMapSimple) equal(other objectImpl) bool {
...
+	var samplesVisited, memUsage, newMemUsage uint64
+	counter := 0
+	totalProps := len(o.data)
+	sampleSize := totalProps / 10
+
+	// grabbing one sample every "sampleSize" to provide consistent
+	// memory usage across function executions
+	for key := range o.data {
+		counter++
+		if counter%sampleSize == 0 {
```
Comment (nathantfrank) [inline object_gomap.go:176]: [minor] if `totalProps` is < 10, then `sampleSize` == 0, when you hit `counter%sampleSize` you will get a `invalid operation: division by zero`.

Maybe only sample if `totalProps` is greater than 10, or maybe if `sampleSize` is 0 then set it to some arbitrary value?

**Review by nathantfrank — CHANGES_REQUESTED** [review]

File: object_gomap.go (hunk 2 above)
Comment (Gabri3l) [reply object_gomap.go:176]: yup I think it's safer to have a guard here, will update with tests, thank you!

**Comment by Gabri3l** [comment]
Took the chance to handle the division by 0 for array and base objects

**Review by nathantfrank — APPROVED** [review]
LGTM

**Pull request merged**